buffer/src/main/java/io/netty/buffer/MiMallocByteBufAllocator.java

```diff
@@ -0,0 +1,2385 @@
+        int mask = alignment - 1;
+        if ((alignment & mask) == 0) {  // If alignment is power of two.
+            return (sz + mask) & ~mask;
+        } else {
+            return ((sz + mask) / alignment) * alignment;
+        }
+    }
+
+    // Round to a good OS allocation size (bounded by max 12.5% waste).
+    static int getGoodOsAllocSize(int size) {
+        int align_size;
+        if (size < 512 * KiB) {
+            align_size = DEFAULT_OS_PAGE_SIZE;
+        } else if (size < 2 * MiB) {
+            align_size = 64 * KiB;
+        } else if (size < 8 * MiB) {
+            align_size = 256 * KiB;
+        } else if (size < 32 * MiB) {
+            align_size = 1 * MiB;
+        } else {
+            align_size = 4 * MiB;
+        }
+        return alignUp(size, align_size);
+    }
+
+    /**
+     * The strategy for how {@link MiMallocByteBufAllocator} should allocate chunk buffers.
+     */
+    interface ChunkAllocator {
+        /**
+         * Allocate a buffer for a chunk. This can be any kind of {@link AbstractByteBuf} implementation.
+         * @param initialCapacity The initial capacity of the returned {@link AbstractByteBuf}.
+         * @param maxCapacity The maximum capacity of the returned {@link AbstractByteBuf}.
+         * @return The buffer that represents the chunk memory.
+         */
+        AbstractByteBuf allocate(int initialCapacity, int maxCapacity);
+    }
+
+    static class MiByteBuf extends AbstractReferenceCountedByteBuf {
+        private int length;
+        private int maxFastCapacity;
+        private AbstractByteBuf rootParent;
+        private int adjustment;
+        private ByteBuffer tmpNioBuf;
+        private boolean hasArray;
+        private boolean hasMemoryAddress;
+        private Block block;
+
+        MiByteBuf() {
+            super(0);
+        }
+
+        void init(Block block, int length, int maxCapacity) {
+            assert block != null;
+            block.nextBlock = null;
+            this.resetRefCnt();
+            this.discardMarks();
+            this.block = block;
+            this.length = length;
+            this.maxFastCapacity = block.blockBytes;
+            this.adjustment = block.blockAdjustment;
+            maxCapacity(maxCapacity);
+            setIndex0(0, 0);
+            this.rootParent = block.page.segment.delegate;
+            this.tmpNioBuf = null;
+            this.hasArray = block.page.segment.delegate.hasArray();
+            this.hasMemoryAddress = block.page.segment.delegate.hasMemoryAddress();
+        }
+
+        @Override
+        protected void deallocate() {
+            assert this.block != null;
+            this.length = 0;
+            this.maxFastCapacity = 0;
+            this.adjustment = 0;
+            this.rootParent = null;
+            this.tmpNioBuf = null;
+            this.hasArray = false;
+            this.hasMemoryAddress = false;
+            MiMallocByteBufAllocator allocator = this.block.page.segment.parent;
```

Comment:
this is performing several data dependent loads (which are still  not covered by Apple's prefetchers AFAIK -> https://news.ycombinator.com/item?id=33474005).
It would be better to save it and performance will further improve